src/main/java/soot/ModulePathSourceLocator.java

```diff
@@ -656,66 +621,62 @@ protected void setupClassProviders() {
+      throw new RuntimeException("Invalid class source type");
+    }
 
-    if (cst == ClassSourceType.directory || cst == ClassSourceType.jrt) {
-
-      FileVisitor<Path> fileVisitor = new FileVisitor<Path>() {
-
-        @Override
-        public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
-          return FileVisitResult.CONTINUE;
-        }
-
-        @Override
-        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
-
-          // Note that some FileSystem implementations used by Path use even on Windows
-          // "/" as path separator. Therefore we have to use the file-system specific path separator
-          // instead of the system specific one (File.separatorChar).
-          String fileName = aPath.relativize(file).toString().replace(file.getFileSystem().getSeparator(), ".");
-
-          if (fileName.endsWith(".class")) {
-            int index = fileName.lastIndexOf(".class");
-            classes.add(fileName.substring(0, index));
-          }
-
-          if (fileName.endsWith(".jimple")) {
-            int index = fileName.lastIndexOf(".jimple");
-            classes.add(fileName.substring(0, index));
-          }
-
-          if (fileName.endsWith(".java")) {
-            int index = fileName.lastIndexOf(".java");
-            classes.add(fileName.substring(0, index));
-          }
+    List<String> classes = new ArrayList<String>();
+    FileVisitor<Path> fileVisitor = new FileVisitor<Path>() {
 
-          return FileVisitResult.CONTINUE;
-        }
-
-        @Override
-        public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
-          return FileVisitResult.CONTINUE;
-        }
+      @Override
+      public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+        return FileVisitResult.CONTINUE;
+      }
 
-        @Override
-        public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
-          return FileVisitResult.CONTINUE;
+      @Override
+      public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+        // Note that some FileSystem implementations used by Path use even on Windows
+        // "/" as path separator. Therefore we have to use the file-system specific path
+        //  separator instead of the system specific one (File.separatorChar).
+        String fileName = aPath.relativize(file).toString().replace(file.getFileSystem().getSeparator(), ".");
+        if (fileName.endsWith(".class")) {
+          classes.add(fileName.substring(0, fileName.lastIndexOf(".class")));
+        } else if (fileName.endsWith(".jimple")) {
+          classes.add(fileName.substring(0, fileName.lastIndexOf(".jimple")));
+        } else if (fileName.endsWith(".java")) {
+          classes.add(fileName.substring(0, fileName.lastIndexOf(".java")));
```

Comment:
if the condition is true we can calculate the end position of the substring to avoid additional search costs of O( suffix.length() ) so we can replace `fileName.lastIndexOf( suffix )` by filename.length()-suffix.length()
```suggestion
        if (fileName.endsWith(".class")) {
          classes.add(fileName.substring(0, fileName.length()-6));
        } else if (fileName.endsWith(".jimple")) {
          classes.add(fileName.substring(0, fileName.length()-7));
        } else if (fileName.endsWith(".java")) {
          classes.add(fileName.substring(0, fileName.length()-5));
```

Reply:
That makes the code harder to understand, do we know this to be a bottleneck?

Reply:
I'll agree.. suggestion updated.